File: jvm/selfie-lib/src/commonMain/kotlin/com/diffplug/selfie/guts/Literals.kt
/*
 * Copyright (C) 2023-2024 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie.guts

import kotlin.math.abs

internal expect fun jreVersion(): Int

enum class Language {
  JAVA,
  JAVA_PRE15,
  KOTLIN,
  GROOVY,
  SCALA;

  companion object {
    fun fromFilename(filename: String): Language {
      return when (filename.substringAfterLast('.')) {
        "java" -> if (jreVersion() < 15) JAVA_PRE15 else JAVA
        "kt" -> KOTLIN
        "groovy",
        "gvy",
        "gy" -> GROOVY
        "scala",
        "sc" -> SCALA
        else -> throw IllegalArgumentException("Unknown language for file $filename")
      }
    }
  }
}

class LiteralValue<T : Any>(val expected: T?, val actual: T, val format: LiteralFormat<T>)

abstract class LiteralFormat<T : Any> {
  internal abstract fun encode(value: T, language: Language): String
  internal abstract fun parse(str: String, language: Language): T
}

private const val MAX_RAW_NUMBER = 1000
private const val PADDING_SIZE = MAX_RAW_NUMBER.toString().length - 1
private fun encodeUnderscores(
    buffer: StringBuilder,
    value: Long,
    language: Language
): StringBuilder {
  return if (value >= MAX_RAW_NUMBER) {
    val mod = value % MAX_RAW_NUMBER
    val leftPadding = PADDING_SIZE - mod.toString().length
    encodeUnderscores(buffer, value / MAX_RAW_NUMBER, language)
    buffer.append("_")
    for (i in leftPadding downTo 1) {
      buffer.append('0')
    }
    buffer.append(mod)
    buffer
  } else if (value < 0) {
    buffer.append('-')
    encodeUnderscores(buffer, abs(value), language)
  } else {
    buffer.append(value)
  }
}

internal object LiteralInt : LiteralFormat<Int>() {
  override fun encode(value: Int, language: Language): String {
    return encodeUnderscores(StringBuilder(), value.toLong(), language).toString()
  }
  override fun parse(str: String, language: Language): Int {
    return str.replace("_", "").toInt()
  }
}

internal object LiteralLong : LiteralFormat<Long>() {
  override fun encode(value: Long, language: Language): String {
    val buffer = encodeUnderscores(StringBuilder(), value, language)
    buffer.append('L')
    return buffer.toString()
  }
  override fun parse(str: String, language: Language): Long {
    var longStr = str.replace("_", "")
    if (longStr.endsWith("L")) {
      longStr = longStr.substring(0, longStr.length - 1)
    }
    return longStr.toLong()
  }
}

private const val TRIPLE_QUOTE = "\"\"\""
private const val KOTLIN_DOLLAR = "\${'\$'}"
private const val KOTLIN_DOLLARQUOTE = "\${'\"'}"

internal object LiteralString : LiteralFormat<String>() {
  override fun encode(value: String, language: Language): String =
      if (value.indexOf('\n') == -1)
          when (language) {
            Language.SCALA, // scala only does $ substitution for s" and f" strings
            Language.JAVA_PRE15,
            Language.GROOVY,
            Language.JAVA -> encodeSingleJava(value)
            Language.KOTLIN -> encodeSingleJavaWithDollars(value)
          }
      else
          when (language) {
            // TODO: support triple-quoted strings in scala
            // https://github.com/diffplug/selfie/issues/106
            Language.SCALA,
            // TODO: support triple-quoted strings in groovy
            // https://github.com/diffplug/selfie/issues/105
            Language.GROOVY,
            Language.JAVA_PRE15 -> encodeSingleJava(value)
            Language.JAVA -> encodeMultiJava(value)
            Language.KOTLIN -> encodeMultiKotlin(value)
          }
  override fun parse(str: String, language: Language): String =
      if (!str.startsWith(TRIPLE_QUOTE))
          when (language) {
            Language.SCALA,
            Language.JAVA_PRE15,
            Language.JAVA -> parseSingleJava(str)
            Language.GROOVY,
            Language.KOTLIN -> parseSingleJavaWithDollars(str)
          }
      else
          when (language) {
            Language.SCALA ->
                throw UnsupportedOperationException(
                    "Selfie doesn't support triple-quoted strings in Scala, yet - help wanted: https://github.com/diffplug/selfie/issues/106")
            Language.GROOVY ->
                throw UnsupportedOperationException(
                    "Selfie doesn't support triple-quoted strings in Groovy, yet - help wanted: https://github.com/diffplug/selfie/issues/105")
            Language.JAVA_PRE15,
            Language.JAVA -> parseMultiJava(str)
            Language.KOTLIN -> parseMultiKotlin(str)
          }
  fun encodeSingleJava(value: String): String = encodeSingleJavaish(value, false)
  fun encodeSingleJavaWithDollars(value: String) = encodeSingleJavaish(value, true)
  private fun encodeSingleJavaish(value: String, escapeDollars: Boolean): String {
    val source = StringBuilder()
    source.append("\"")
    for (char in value) {
      when (char) {
        '\b' -> source.append("\\b")
        '\n' -> source.append("\\n")
        '\r' -> source.append("\\r")
        '\t' -> source.append("\\t")
        '\"' -> source.append("\\\"")
        '\\' -> source.append("\\\\")
        '$' -> if (escapeDollars) source.append(KOTLIN_DOLLAR) else source.append('$')
        else ->
            if (isControlChar(char)) {
              source.append("\\u")
              source.append(char.code.toString(16).padStart(4, '0'))
            } else {
              source.append(char)
            }
      }
    }
    source.append("\"")
    return source.toString()
  }
  private fun isControlChar(c: Char): Boolean {
    return c in '\u0000'..'\u001F' || c == '\u007F'
  }
  fun parseSingleJava(sourceWithQuotes: String) = parseSingleJavaish(sourceWithQuotes, false)
  fun parseSingleJavaWithDollars(sourceWithQuotes: String) =
      parseSingleJavaish(sourceWithQuotes, true)
  private fun parseSingleJavaish(sourceWithQuotes: String, removeDollars: Boolean): String {
    check(sourceWithQuotes.startsWith('"'))
    check(sourceWithQuotes.endsWith('"'))
    val source = sourceWithQuotes.substring(1, sourceWithQuotes.length - 1)
    val toUnescape = if (removeDollars) inlineDollars(source) else source
    return unescapeJava(toUnescape)
  }
  fun encodeMultiKotlin(arg: String): String {
    val escapeDollars = arg.replace("$", KOTLIN_DOLLAR)
    val escapeTripleQuotes =
        escapeDollars.replace(
            TRIPLE_QUOTE, "$KOTLIN_DOLLARQUOTE$KOTLIN_DOLLARQUOTE$KOTLIN_DOLLARQUOTE")
    val protectWhitespace =
        escapeTripleQuotes.lines().joinToString("\n") { line ->
          val protectTrailingWhitespace =
              if (line.endsWith(" ")) {
                line.dropLast(1) + "\${' '}"
              } else if (line.endsWith("\t")) {
                line.dropLast(1) + "\${'\\t'}"
              } else line
          val protectLeadingWhitespace =
              if (protectTrailingWhitespace.startsWith(" ")) {
                "\${' '}" + protectTrailingWhitespace.drop(1)
              } else if (protectTrailingWhitespace.startsWith("\t")) {
                "\${'\\t'}" + protectTrailingWhitespace.drop(1)
              } else protectTrailingWhitespace
          protectLeadingWhitespace
        }
    return "$TRIPLE_QUOTE$protectWhitespace$TRIPLE_QUOTE"
  }
  fun encodeMultiJava(arg: String): String {
    val escapeBackslashes = arg.replace("\\", "\\\\")
    val escapeTripleQuotes = escapeBackslashes.replace(TRIPLE_QUOTE, "\\\"\\\"\\\"")
    val protectWhitespace =
        escapeTripleQuotes.lines().joinToString("\n") { line ->
          val protectTrailingWhitespace =
              if (line.endsWith(" ")) {
                line.dropLast(1) + "\\s"
              } else if (line.endsWith("\t")) {
                line.dropLast(1) + "\\t"
              } else line
          val protectLeadingWhitespace =
              if (protectTrailingWhitespace.startsWith(" ")) {
                "\\s" + protectTrailingWhitespace.drop(1)
              } else if (protectTrailingWhitespace.startsWith("\t")) {
                "\\t" + protectTrailingWhitespace.drop(1)
              } else protectTrailingWhitespace
          protectLeadingWhitespace
        }
    return "$TRIPLE_QUOTE\n$protectWhitespace$TRIPLE_QUOTE"
  }
  private val charLiteralRegex = """\$\{'(\\?.)'\}""".toRegex()
  private fun inlineDollars(source: String): String {
    if (source.indexOf('$') == -1) {
      return source
    }
    return charLiteralRegex.replace(source) { matchResult ->
      val charLiteral = matchResult.groupValues[1]
      when {
        charLiteral.length == 1 -> charLiteral
        charLiteral.length == 2 && charLiteral[0] == '\\' ->
            when (charLiteral[1]) {
              't' -> "\t"
              'b' -> "\b"
              'n' -> "\n"
              'r' -> "\r"
              '\'' -> "'"
              '\\' -> "\\"
              '$' -> "$"
              else -> charLiteral
            }
        else -> throw IllegalArgumentException("Unknown character literal $charLiteral")
      }
    }
  }
  private fun unescapeJava(source: String): String {
    val firstEscape = source.indexOf('\\')
    if (firstEscape == -1) {
      return source
    }
    val value = StringBuilder()
    value.append(source.substring(0, firstEscape))
    var i = firstEscape
    while (i < source.length) {
      var c = source[i]
      if (c == '\\') {
        i++
        c = source[i]
        when (c) {
          '\"' -> value.append('\"')
          '\\' -> value.append('\\')
          'b' -> value.append('\b')
          'f' -> value.append('\u000c')
          'n' -> value.append('\n')
          'r' -> value.append('\r')
          's' -> value.append(' ')
          't' -> value.append('\t')
          'u' -> {
            val code = source.substring(i + 1, i + 5).toInt(16)
            value.append(code.toChar())
            i += 4
          }
          else -> throw IllegalArgumentException("Unknown escape sequence $c")
        }
      } else {
        value.append(c)
      }
      i++
    }
    return value.toString()
  }
  fun parseMultiJava(sourceWithQuotes: String): String {
    check(sourceWithQuotes.startsWith("$TRIPLE_QUOTE\n"))
    check(sourceWithQuotes.endsWith(TRIPLE_QUOTE))
    val source =
        sourceWithQuotes.substring(
            TRIPLE_QUOTE.length + 1, sourceWithQuotes.length - TRIPLE_QUOTE.length)
    val lines = source.lines()
    val commonPrefix =
        lines
            .mapNotNull { line ->
              if (line.isNotBlank()) line.takeWhile { it.isWhitespace() } else null
            }
            .minOrNull() ?: ""
    return lines.joinToString("\n") { line ->
      if (line.isBlank()) {
        ""
      } else {
        val removedPrefix = if (commonPrefix.isEmpty()) line else line.removePrefix(commonPrefix)
        val removeTrailingWhitespace = removedPrefix.trimEnd()
        val handleEscapeSequences = unescapeJava(removeTrailingWhitespace)
        handleEscapeSequences
      }
    }
  }
  fun parseMultiKotlin(sourceWithQuotes: String): String {
    check(sourceWithQuotes.startsWith(TRIPLE_QUOTE))
    check(sourceWithQuotes.endsWith(TRIPLE_QUOTE))
    val source =
        sourceWithQuotes.substring(
            TRIPLE_QUOTE.length, sourceWithQuotes.length - TRIPLE_QUOTE.length)
    return inlineDollars(source)
  }
}

internal object LiteralBoolean : LiteralFormat<Boolean>() {
  override fun encode(value: Boolean, language: Language): String {
    return value.toString()
  }
  override fun parse(str: String, language: Language): Boolean {
    return str.toBooleanStrict()
  }
}

internal object DiskSnapshotTodo : LiteralFormat<Unit>() {
  override fun encode(value: Unit, language: Language) = throw UnsupportedOperationException()
  override fun parse(str: String, language: Language) = throw UnsupportedOperationException()
  fun createLiteral() = LiteralValue(null, Unit, DiskSnapshotTodo)
}

internal object ToBeFileTodo : LiteralFormat<Unit>() {
  override fun encode(value: Unit, language: Language) = throw UnsupportedOperationException()
  override fun parse(str: String, language: Language) = throw UnsupportedOperationException()
  fun createLiteral() = LiteralValue(null, Unit, ToBeFileTodo)
}

File: python/selfie-lib/selfie_lib/Literals.py
from enum import Enum, auto
from typing import Protocol, TypeVar
from abc import abstractmethod
from .EscapeLeadingWhitespace import EscapeLeadingWhitespace
import io
import re

T = TypeVar("T")


class Language(Enum):
    PYTHON = auto()

    @classmethod
    def from_filename(cls, filename: str) -> "Language":
        extension = filename.rsplit(".", 1)[-1]
        if extension == "py":
            return cls.PYTHON
        else:
            raise ValueError(f"Unknown language for file {filename}")


class LiteralValue:
    def __init__(self, expected: T | None, actual: T, format: "LiteralFormat") -> None:
        self.expected = expected
        self.actual = actual
        self.format = format


class LiteralFormat(Protocol[T]):
    @abstractmethod
    def encode(
        self, value: T, language: Language, encoding_policy: "EscapeLeadingWhitespace"
    ) -> str:
        raise NotImplementedError("Subclasses must implement the encode method")

    @abstractmethod
    def parse(self, string: str, language: Language) -> T:
        raise NotImplementedError("Subclasses must implement the parse method")


MAX_RAW_NUMBER = 1000
PADDING_SIZE = len(str(MAX_RAW_NUMBER)) - 1


class LiteralInt(LiteralFormat[int]):
    def _encode_underscores(
        self, buffer: io.StringIO, value: int, language: Language
    ) -> io.StringIO:
        if value >= MAX_RAW_NUMBER:
            mod = value % MAX_RAW_NUMBER
            left_padding = PADDING_SIZE - len(str(mod))
            self._encode_underscores(buffer, value // MAX_RAW_NUMBER, language)
            buffer.write("_")
            buffer.write("0" * left_padding)
            buffer.write(str(mod))
            return buffer
        elif value < 0:
            buffer.write("-")
            self._encode_underscores(buffer, abs(value), language)
            return buffer
        else:
            buffer.write(str(value))
            return buffer

    def encode(
        self, value: int, language: Language, encoding_policy: EscapeLeadingWhitespace
    ) -> str:
        return self._encode_underscores(io.StringIO(), value, language).getvalue()

    def parse(self, string: str, language: Language) -> int:
        return int(string.replace("_", ""))


TRIPLE_QUOTE = '"""'


class LiteralString(LiteralFormat[str]):
    def encode(
        self, value: str, language: Language, encoding_policy: EscapeLeadingWhitespace
    ) -> str:
        if language == Language.PYTHON:
            if "/n" not in value:
                return self._encodeSinglePython(value)
            else:
                return self.encodeMultiPython(value, encoding_policy)
        else:
            raise NotImplementedError(
                "Encoding for language {} is not implemented.".format(language)
            )

    def parse(self, string: str, language: Language) -> str:
        if language == Language.PYTHON:
            if not string.startswith(TRIPLE_QUOTE):
                return self._parseSinglePython(string)
            else:
                return self.parseMultiPython(string)
        else:
            raise NotImplementedError(
                "Encoding for language {} is not implemented.".format(language)
            )

    def _encodeSinglePython(self, value: str) -> str:
        source = io.StringIO()
        source.write('"')
        for char in value:
            if char == "\b":
                source.write("\\b")
            elif char == "\n":
                source.write("\\n")
            elif char == "\r":
                source.write("\\r")
            elif char == "\t":
                source.write("\\t")
            elif char == '"':
                source.write('\\"')
            elif char == "\\":
                source.write("\\\\")
            elif self._is_control_char(char):
                source.write("\\u" + str(ord(char)).zfill(4))
            else:
                source.write(char)
        source.write('"')
        return source.getvalue()

    def _is_control_char(self, c: str) -> bool:
        return c in "\u0000\u001f" or c == "\u007f"

    # combined logic from parseSingleJava and parseSingleJavaish
    def _parseSinglePython(self, source_with_quotes: str) -> str:
        assert source_with_quotes.startswith('"')
        assert source_with_quotes.endswith('"')
        source = source_with_quotes[1:-1]
        to_unescape = self.inline_backslashes(source)  # changed from inline_dollar
        return self._unescape_python(to_unescape)

    def encodeMultiPython(
        self, arg: str, escape_leading_whitespace: EscapeLeadingWhitespace
    ) -> str:
        escape_backslashes = arg.replace("\\", "\\\\")
        escape_triple_quotes = escape_backslashes.replace(TRIPLE_QUOTE, '\\"\\"\\"')

        def protect_trailing_whitespace(line):
            if line.endswith(" "):
                return line[:-1] + "\\u0020"
            elif line.endswith("\t"):
                return line[:-1] + "\\t"
            else:
                return line

        lines = escape_triple_quotes.splitlines()
        protect_whitespace = "\n".join(
            escape_leading_whitespace.escape_line(
                protect_trailing_whitespace(line), "\\u0020", "\\t"
            )
            for line in lines
        )

        return f"{TRIPLE_QUOTE}\n{protect_whitespace}{TRIPLE_QUOTE}"

    _char_literal_pattern = re.compile(r"""\{'(\\?.)'\}""")

    def inline_backslashes(self, source: str) -> str:
        def replace_char(char_literal: str) -> str:
            if len(char_literal) == 1:
                return char_literal
            elif len(char_literal) == 2 and char_literal[0] == "\\":
                if char_literal[1] == "t":
                    return "\t"
                elif char_literal[1] == "b":
                    return "\b"
                elif char_literal[1] == "n":
                    return "\n"
                elif char_literal[1] == "r":
                    return "\r"
                elif char_literal[1] == "'":
                    return "'"
                elif char_literal[1] == "\\":
                    return "\\"
                else:
                    raise ValueError(f"Unknown character literal {char_literal}")
            else:
                raise ValueError(f"Unknown character literal {char_literal}")

        return self._char_literal_pattern.sub(
            lambda match: replace_char(match.group(1)), source
        )

    def _unescape_python(self, source: str) -> str:
        value = io.StringIO()
        i = 0
        while i < len(source):
            c = source[i]
            if c == "\\":
                i += 1
                c = source[i]
                if c == '"':
                    value.write('"')
                elif c == "\\":
                    value.write("\\")
                elif c == "b":
                    value.write("\b")
                elif c == "f":
                    value.write("\f")
                elif c == "n":
                    value.write("\n")
                elif c == "r":
                    value.write("\r")
                elif c == "s":
                    value.write(" ")
                elif c == "t":
                    value.write("\t")
                elif c == "u":
                    code = int(source[i + 1 : i + 5], 16)
                    value.write(chr(code))
                    i += 4
                else:
                    raise ValueError(f"Unknown escape sequence {c}")
            else:
                value.write(c)
            i += 1
        return value.getvalue()

    def parseMultiPython(self, source_with_quotes: str) -> str:
        assert source_with_quotes.startswith(TRIPLE_QUOTE + "\n")
        assert source_with_quotes.endswith(TRIPLE_QUOTE)

        source = source_with_quotes[len(TRIPLE_QUOTE) + 1 : -len(TRIPLE_QUOTE)]
        lines = source.split("\n")

        common_prefix = min(
            (line[: len(line) - len(line.lstrip())] for line in lines if line.strip()),
            default="",
        )

        def remove_common_prefix(line: str) -> str:
            return line[len(common_prefix) :] if common_prefix else line

        def handle_escape_sequences(line: str) -> str:
            return self._unescape_python(line.rstrip())

        return "\n".join(
            handle_escape_sequences(remove_common_prefix(line))
            for line in lines
            if line.strip()
        )


class LiteralBoolean(LiteralFormat[bool]):
    def encode(
        self, value: bool, language: Language, encoding_policy: EscapeLeadingWhitespace
    ) -> str:
        return str(value)

    def __to_boolean_strict(self, string: str) -> bool:
        if string.lower() == "true":
            return True
        elif string.lower() == "false":
            return False
        else:
            raise ValueError("String is not a valid boolean representation: " + string)

    def parse(self, string: str, language: Language) -> bool:
        return self.__to_boolean_strict(string)
